src/backend/distributed/operations/shard_split.c

```diff
@@ -0,0 +1,920 @@
+
+		/* NodeId in Citus are unsigned and range from [1, 4294967296]. */
+		if (nodeIdValue < 1 || workerNode == NULL)
+		{
+			ereport(ERROR, (errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+							errmsg("Invalid Node Id '%u'.", nodeIdValue)));
+		}
+
+		workersForPlacementList =
+			lappend(workersForPlacementList, (void *) workerNode);
+	}
+
+	if (splitMode == BLOCKING_SPLIT)
+	{
+		EnsureReferenceTablesExistOnAllNodesExtended(TRANSFER_MODE_BLOCK_WRITES);
+		BlockingShardSplit(
+			splitOperation,
+			shardIntervalToSplit,
+			shardSplitPointsList,
+			workersForPlacementList);
+	}
+	else
+	{
+		/* we only support blocking shard split in this code path for now. */
+		ereport(ERROR, (errmsg("Invalid split mode value %d.", splitMode)));
+	}
+}
+
+
+/*
+ * SplitShard API to split a given shard (or shard group) in blocking fashion
+ * based on specified split points to a set of destination nodes.
+ * 'splitOperation'             : Customer operation that triggered split.
+ * 'shardIntervalToSplit'       : Source shard interval to be split.
+ * 'shardSplitPointsList'		: Split Points list for the source 'shardInterval'.
+ * 'workersForPlacementList'	: Placement list corresponding to split children.
+ */
+static void
+BlockingShardSplit(SplitOperation splitOperation,
+				   ShardInterval *shardIntervalToSplit,
+				   List *shardSplitPointsList,
+				   List *workersForPlacementList)
+{
+	List *sourceColocatedShardIntervalList = ColocatedShardIntervalList(
+		shardIntervalToSplit);
+
+	BlockWritesToShardList(sourceColocatedShardIntervalList);
+
+	/* First create shard interval metadata for split children */
+	List *shardGroupSplitIntervalListList = CreateSplitIntervalsForShardGroup(
+		sourceColocatedShardIntervalList,
+		shardSplitPointsList);
+
+	/* Only single placement allowed (already validated RelationReplicationFactor = 1) */
+	List *sourcePlacementList = ActiveShardPlacementList(shardIntervalToSplit->shardId);
+	Assert(sourcePlacementList->length == 1);
+	ShardPlacement *sourceShardPlacement = (ShardPlacement *) linitial(
+		sourcePlacementList);
+	WorkerNode *sourceShardToCopyNode = FindNodeWithNodeId(sourceShardPlacement->nodeId,
+														   false /* missingOk */);
+
+	PG_TRY();
+	{
+		/*
+		 * Physically create split children, perform split copy and create auxillary structures.
+		 * This includes: indexes, replicaIdentity. triggers and statistics.
+		 * Foreign key constraints are created after Metadata changes (see CreateForeignKeyConstraints).
+		 */
+		CreateSplitShardsForShardGroup(
+			sourceShardToCopyNode,
+			sourceColocatedShardIntervalList,
+			shardGroupSplitIntervalListList,
+			workersForPlacementList);
+
+		/*
+		 * Drop old shards and delete related metadata. Have to do that before
+		 * creating the new shard metadata, because there's cross-checks
+		 * preventing inconsistent metadata (like overlapping shards).
+		 */
+		DropShardList(sourceColocatedShardIntervalList);
```

Comment:
One of the side-effects of this function call will be rather scary error messages for the user about shards no longer existing.

session 1:
```sql
begin;
SELECT citus_split_shard_by_split_points(
102039 ,
ARRAY['2093265920'],
ARRAY[2, 3]);
```
session 2:
```sql
select count(*) from test;
```

session 1:
```sql
end;
```

session 2:
```sql
ERROR:  relation "public.test_102039" does not exist
CONTEXT:  while executing command on localhost:1402
```

In case of writes, we get the slightly less scary:
```sql
ERROR:  could not find valid entry for shard 102039
```

The solution would be to implement deferred drop, but we'll need to be super careful about how we handle writes, since shard placement selection is deferred until after obtaining shard-level locks, but shard selection is not.

I think we need a comprehensive redesign of deferred drop & general clean-up, because we also have new requirements from the shard move side (e.g. doing moves in parallel). We also need to handle failure scenarios where splits are left behind, since there is currently no guarantee that they will ever be cleaned up.

For the current PR, we can probably live with these errors.

Reply:
Thanks for the writeup and explaining the problem. I will think more about this problem and propose a design.